### ISSUE TITLE
why mwer use stop gradient?

### ISSUE THREAD

**Issue by Mddct** [TeaPoly/CTC-MWER#2]
why mwer use stop gradient? just a regularization?

**Comment by Mddct** [comment]
> why mwer use stop gradient? just a regularization?

May be Variance reduction

**Comment by leixiaoning** [comment]
i find tf ctc beam search will loss the gradients

**Comment by TeaPoly** [comment]
> i find tf ctc beam search will loss the gradients

Beam search is just to find candidate paths, gradient is not required in beam search. Gradients are pushed back to `logit` weight since there are probability `P` which is computed from `logit` ​​as input to MWER loss. NBEST path from CTC Beam search can actually be generated offline to speed up training.

![](https://user-images.githubusercontent.com/3815778/206600760-ea1927ea-e479-43b0-8aa4-14c10ede7511.png)
